### PR TITLE
ARROW-14441: [R] Add our philosophy to the dev vignette

### PR DESCRIPTION
**Pull request by thisisnic**
* separates some of the developer docs into separate files (content remains unchanged)
* updates the original dev docs page to point to these other pages and discuss our philosophy when implementing bindings

### CODE REVIEW

File: r/vignettes/developing.Rmd
Comment (thisisnic) [inline r/vignettes/developing.Rmd:68]: I think this vignette needs more after here but I don't know exactly what.  Maybe something on writing bindings between compute kernels and R functions? Or is that a bit too specific?

Comment (wjones127) [reply r/vignettes/developing.Rmd:68]: I think an example of a compute kernel binding would be good. We could use that as an example to further explain the above points.

I'd also suggest linking to recent PRs that may serve an examples. I find a reference PR particularly helpful in reminding me which files I might need to modify.

**Review by wjones127 — CHANGES_REQUESTED** [review]
Thanks for organizing this! These developer docs are already quite nice and I'm glad to see them further enhanced.

Saw one broken link, and then a few other suggestions.

File: r/vignettes/developing.Rmd
Comment (thisisnic) [reply r/vignettes/developing.Rmd:47]: Unfortunately not - when the package is built, everything in the `vignettes` directory is distributed with the package as a HTML document, whereas everything in any subdirectories is only displayed on the `pkgdown` site.  This means that the relative links wouldn't work for anyone viewing this vignette locally.

File: r/vignettes/developers/setup.Rmd
Comment (thisisnic) [reply r/vignettes/developers/setup.Rmd:297]: Good catch!

File: r/vignettes/developing.Rmd
Comment (thisisnic) [reply r/vignettes/developing.Rmd:55]: Nice, I like this rephrasing

File: r/vignettes/developing.Rmd
Comment (thisisnic) [reply r/vignettes/developing.Rmd:68]: Great point, agreed on the reference PR.  I'm also gonna tag some of the wider R dev team to pitch in on this, as I can't help but feeling there's a bit more to discuss on what docs we distribute with the package vs. what docs we just have on the pkgdown site.

Comment (thisisnic) [reply r/vignettes/developing.Rmd:68]: @jonkeane What are your thoughts on this?

Comment (jonkeane) [reply r/vignettes/developing.Rmd:68]: Personally, I think of the pkgdown site as *the* canonical documentation, especially for vignettes. So I don't have a strong opinion / I'm not worried about not including it inside of / distributed along side the package.

As for examples: I think that would be great. We can link to PRs (though if the link goes to CRAN it's susceptible to a redirect/rot that will anger CRAN — we should and do check for that, but just a reminder). Though sometimes when writing examples I find it a little bit easier to make a dedicated example that has lots of extra comments/commentary/possibly even glosses over some of the reality involved with implementing them.

File: r/vignettes/developers/setup.Rmd
Comment (jonkeane) [inline r/vignettes/developers/setup.Rmd:56]: While we're in here, could we fold in the spirit of https://issues.apache.org/jira/browse/ARROW-14371?

Specifically:

* A note that the brew-based method and the build-your-own methods are incompatible (mostly because of ARROW_HOME, but folks using brew shouldn't need to even think about that, so we should be careful how we phrase this)
* A note about confirming that brew install apache-arrow --HEAD completes successfully and how to confirm it's being picked up in the install process
* This might be part of the point above, but a description of what the difference/meaning of the following are: *** Using Homebrew ${PKG_BREW_NAME}, *** Arrow C++ libraries found via pkg-config, any mention of autobrew. Specifically, if one is trying to use homebrew for development, only the first one is ok, if someone sees something else that means that something isn't quite right.

If this is too much or you don't want to extend scope, that's totally fine!

File: r/vignettes/developers/workflow.Rmd
Comment (jonkeane) [inline r/vignettes/developers/workflow.Rmd:18]: Do we still need this? I know we are waiting on a release, but can't remember if roxygen2 has had one recently or not

Comment (thisisnic) [reply r/vignettes/developers/workflow.Rmd:18]: Yeah, there's a recent update, I'll update this in once I sort the merge conflict.

File: r/vignettes/developers/setup.Rmd
Comment (thisisnic) [reply r/vignettes/developers/setup.Rmd:43]: Will leave this for now as I don't fully understand all of these things, so am leaving it for another PR

**Comment by thisisnic** [comment]
@jonkeane and @wjones127 - just rebased this so now the setup instructions contain the excellent changes made by @wjones127 including all that Windows content.  Just thinking - I am in the process of writing up some content on writing bindings, but it's a bigger piece of work.  Any objections to me doing that in a separate follow-up ticket?

File: r/vignettes/developing.Rmd
Comment (wjones127) [reply r/vignettes/developing.Rmd:16]: Ohhh. Does this mean we are making parts of these developer docs not available offline?

**Review by wjones127 — APPROVED** [review]
Yeah I think doing the example as a follow-up makes sense.

File: r/vignettes/developing.Rmd
Comment (thisisnic) [reply r/vignettes/developing.Rmd:16]: Yeah, there's no real need to distribute them with the package given that most people read this content via the pkgdown site anyway.

**Comment by thisisnic** [comment]
I'll cover the bindings stuff in https://issues.apache.org/jira/browse/ARROW-14757

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-devdocs

**Comment by jonkeane** [comment]
This looks good.  Just to confirm, the contents of r/vignettes/developers/workflow.Rmd and r/vignettes/developers/setup.Rmd  were all just copy/pasted and don't have substantive edits do they? If they do, would you mind pointing them out so we can make sure to take a look at them?

Also I presume you've run pkgdown locally to see that this all looks ok. NBD if it isn't perfect on the first merge, we've also got lots of time to look at it on the dev docs site before the 7.0.0 release

**Review by jonkeane — APPROVED** [review]
LGTM, pending the crossbow job. If that passes feel free to merge

**Comment by thisisnic** [comment]
> This looks good. Just to confirm, the contents of r/vignettes/developers/workflow.Rmd and r/vignettes/developers/setup.Rmd were all just copy/pasted and don't have substantive edits do they? If they do, would you mind pointing them out so we can make sure to take a look at them?
> 
> Also I presume you've run pkgdown locally to see that this all looks ok. NBD if it isn't perfect on the first merge, we've also got lots of time to look at it on the dev docs site before the 7.0.0 release

Yeah, no substantive edits and will make sure I run pkgdown to check it all before merging.

**Pull request closed without merging**
